### DISCUSSION TITLE
large systems

### DISCUSSION THREAD

**Discussion by willphy123** [ttadano/alamode#76] [Q&A]
Dear Dr. Tadano,
        I have a large system that has ~350 atoms. I created a 2*2*1 supercell to produce displacement patterns. Because the system has no symmetry, there are a lot of displacements. I want to ask if alm can do parallelization to create displacement patterns? The calculations are very slow. Can you give me advice to accelerate creating them?
Thanks
Will

**Comment by ttadano** [comment]
Do you want to parallielize the calculations of atomic forces for the 221 structures? If so, that can be done by submitting several jobs to your supercomputer nodes, each of which works on a different subset of the structures.

**Reply by willphy123** [reply]
Thanks for your quick reply. No, I'm not calculating forces for 221 structures. Instead, I created a 2 by 2 by 1 supercell, not 221 structures. Actually, I am at the very beginning of obtaining displacement patterns. The process of creating displacement patterns is very slow. It is still not finished after two hours. 
So I want to ask if alm can do parallelization to obtain displacement patterns. 
<img width="507" alt="Screen Shot 2022-06-29 at 5 06 05 PM" src="https://user-images.githubusercontent.com/104464250/176565835-b6a0dcd9-7de5-4f9a-b1f5-34ea01119833.png">

**Reply by ttadano** [reply]
ALM is partly parallelized using OpenMP. If you are using a multi- or many-core CPU and the binary is compiled with `-fopenmp` or `-qopenmp` option, the OpenMP parallelization is activated by default. Please see the information in the header part of the log file.

If OpenMP parallelization is already active, you may need to reduce the cutoff radius or use a faster CPU. Since your primitive cell is already fairly large, I think you don't need to use a 2x2x1 supercell; the primitive cell calculation would be sufficient, I guess.

**Reply by willphy123** [reply]
OK. I will try the primitive cell. 

Output:
___________________
+-----------------------------------------------------------------+
 +                         Program ALM                             +
 +                             Ver.  1.4.1                         +
 +-----------------------------------------------------------------+

 +                         Program ALM                             +
 +                             Ver.  1.4.1                         +
 +-----------------------------------------------------------------+

 +-----------------------------------------------------------------+
 +                         Program ALM                             +
 +                             Ver.  1.4.1                         +
 +-----------------------------------------------------------------+

 +-----------------------------------------------------------------+
 +                         Program ALM                             +
 +                             Ver.  1.4.1                         +
 +-----------------------------------------------------------------+

 Number of OpenMP threads = 28

 Number of OpenMP threads = 28

 Job started at Thu Jun 30 05:47:33 2022

 Job started at Thu Jun 30 05:47:33 2022

 Number of OpenMP threads = 28


 Input variables:
 -------------------------------------------------------------------
 General:
  PREFIX = c_221
  MODE = suggest
  NAT = 1456; NKD = 1
  PRINTSYM = 0; TOLERANCE = 0.001
  KD =    C
  FCSYM_BASIS = Lattice
  PERIODIC =   1  1  1
  MAGMOM = 
  FCS_ALAMODE = 1;  NMAXSAVE = 5
  FC3_SHENGBTE = 0
  FC2_QEFC = 0
  HESSIAN = 0
  FC_ZERO_THR = 0

 Interaction:
  NORDER = 1
  NBODY =   2

  DBASIS = CART

 -------------------------------------------------------------------


 Input variables:
 -------------------------------------------------------------------
 General:
  PREFIX = c_221
  MODE = suggest
  NAT = 1456; NKD = 1
  PRINTSYM = 0; TOLERANCE = 0.001
  KD =    C
  FCSYM_BASIS = Lattice
  PERIODIC =   1  1  1
  MAGMOM = 
  FCS_ALAMODE = 1;  NMAXSAVE = 5
  FC3_SHENGBTE = 0
  FC2_QEFC = 0
  HESSIAN = 0
  FC_ZERO_THR = 0

 Interaction:
  NORDER = 1
  NBODY =   2

  DBASIS = CART
-------------------------------------------------

**Reply by LePingKYXK** [reply]
Hi Will,
May I ask whether the parallelization calculation could be run if I install the libraries via conda? Or is the parallelization performed only when libraries are installed from the source?

**Comment by LePingKYXK** [comment]
> ALM is partly parallelized using OpenMP. If you are using a multi- or many-core CPU and the binary is compiled with `-fopenmp` or `-qopenmp` option, the OpenMP parallelization is activated by default. Please see the information in the header part of the log file.
> 
> If OpenMP parallelization is already active, you may need to reduce the cutoff radius or use a faster CPU. Since your primitive cell is already fairly large, I think you don't need to use a 2x2x1 supercell; the primitive cell calculation would be sufficient, I guess.

Dear Dr. Tadano,
As you mentioned that the compile options `-fopenmp` or `-qopenmp`, is that mean the parallelization calculation could be run only if I compile the libraries (the Boost C++, Eigen3, and spglib libraries) from the source and compile `alamode` through `makefile`?

Can I run parallel calculations if I install the libraries via `conda` and then compile `alamode` via `makefile`?

By the way, the link of `spglib` has changed to https://github.com/spglib/spglib. The link (https://alamode.readthedocs.io/en/latest/install.html#step-1-install-all-required-libraries) needs to be updated.

**Reply by ttadano** [reply]
The OpenMP parallelization will also be enabled by the conda build if the compiler supports OpenMP. Please check if the CMake output contains the following line:

```
-- Found OpenMP: TRUE (found version "4.5")
```

If the flag is TRUE, OpenMP is enabled.
